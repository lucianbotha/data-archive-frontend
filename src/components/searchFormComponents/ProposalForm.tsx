import * as React from "react";
import { IGeneral } from "../../utils/ObservationQueryParameters";
import { MainGrid, SubGrid } from "../basicComponents/Grids";
import InputField from "../basicComponents/InputField";

class ProposalForm extends React.Component<
  { proposal: IGeneral; onChange: (value: IGeneral) => void },
  any
> {
  change = (e: React.FormEvent<HTMLInputElement>) => {
    const name = e.currentTarget.name;
    const value = e.currentTarget.value;
    this.props.onChange({
      ...this.props.proposal,
      [name]: value
    });
  };
  render() {
    const {
      errors,
      principalInvestigator,
      observationNight,
      proposalCode,
      proposalTitle
    } = this.props.proposal;
    return (
      <>
        <MainGrid>
          <SubGrid>
            <p>Proposal Code</p>
            <InputField
              error={errors.proposalCode}
              name={"proposalCode"}
              onChange={this.change}
              value={proposalCode}
            />
          </SubGrid>
          <SubGrid>
            <p>Principal investigator</p>
            <InputField
              name={"principalInvestigator"}
              value={principalInvestigator || ""}
              error={errors.principalInvestigator || ""}
              onChange={this.change}
            />
          </SubGrid>
        </MainGrid>

        <MainGrid>
          <SubGrid>
            <p>Proposal title</p>
            <InputField
              name={"proposalTitle"}
              value={proposalTitle}
              error={errors.proposalTitle}
              onChange={this.change}
            />
          </SubGrid>
          <SubGrid>
            <p>Observation night</p>
            <InputField
              name={"observationNight"}
              value={observationNight}
              error={errors.observationNight}
              onChange={this.change}
              placeholder="dd-mm-yyyy"
            />
          </SubGrid>
        </MainGrid>
      </>
    );
  }
}
export default ProposalForm;
