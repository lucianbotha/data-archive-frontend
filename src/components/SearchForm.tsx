import * as React from "react";
import {
  CalibrationType,
  IGeneral,
  ISearchFormState,
  ITarget,
  ITelescope
} from "../utils/ObservationQueryParameters";
import {
  ButtonGrid,
  DataGrid,
  ParentGrid,
  ProposalGrid,
  Spinner,
  TargetGrid,
  TelescopeGrid
} from "./basicComponents/Grids";
import DataForm from "./searchFormComponents/DataForm";
import ProposalForm, {
  validatedProposal
} from "./searchFormComponents/ProposalForm";
import TargetForm, { validatedTarget } from "./searchFormComponents/TargetForm";
import TelescopeForm, {
  validatedTelescope
} from "./searchFormComponents/TelescopeForm";

/**
 * A form for defining search parameters for an observation search, and for
 * initiating the search.
 */
class SearchForm extends React.Component<{}, ISearchFormState> {
  public state: ISearchFormState = {
    general: { calibrations: new Set<CalibrationType>(), errors: {} },
    loading: false,
    target: {
      errors: {},
      resolver: "Simbad",
      searchConeRadiusUnits: "arcseconds"
    }
  };

  /**
   * Handle changes of telescope-related parameters.
   */
  public telescopeChange = (value: ITelescope) => {
    const newState = {
      ...this.state,
      telescope: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of target-related parameters.
   */
  public targetChange = (value: ITarget) => {
    const newState = {
      ...this.state,
      target: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Handle changes of general parameters.
   */
  public generalChange = (value: IGeneral) => {
    const newState = {
      ...this.state,
      general: {
        ...value
      }
    };
    this.setState(() => newState);
  };

  /**
   * Perform an observation with the currently selected search parameters.
   */
  public searchArchive = async () => {
    this.setState(() => ({
      ...this.state,
      loading: true
    }));

    // Add errors to the search parameter details
    const target = await validatedTarget(this.state.target);
    const general = validatedProposal(this.state.general);
    const telescope = validatedTelescope(this.state.telescope);

    this.setState(() => ({
      ...this.state,
      general,
      loading: false,
      target,
      telescope
    }));
  };

  public render() {
    const { target, general, telescope, loading } = this.state;

    return (
      <>
        {loading && (
          <div className="dimScreen">
            <Spinner />
          </div>
        )}
        <ParentGrid>
          <TargetGrid>
            <TargetForm target={target} onChange={this.targetChange} />
          </TargetGrid>
          <ProposalGrid>
            <ProposalForm proposal={general} onChange={this.generalChange} />
          </ProposalGrid>
          <TelescopeGrid>
            <TelescopeForm
              telescope={telescope}
              onChange={this.telescopeChange}
            />
          </TelescopeGrid>
          <DataGrid>
            <DataForm general={general} onChange={this.generalChange} />
          </DataGrid>
          <ButtonGrid>
            <input
              className="button is-primary"
              type="button"
              value="Search"
              onClick={this.searchArchive}
            />
          </ButtonGrid>
        </ParentGrid>
      </>
    );
  }
}

export default SearchForm;
