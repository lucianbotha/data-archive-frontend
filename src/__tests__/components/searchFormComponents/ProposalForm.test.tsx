import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import SearchForm from "../../../components/SearchForm";
import ProposalForm from "../../../components/searchFormComponents/ProposalForm";

const onChange = jest.fn();

describe("Proposal Form", () => {
  const wrapper = mount(
    <ProposalForm proposal={{ errors: {} }} onChange={onChange} />
  );
  it("should render correctly", () => {
    expect(wrapper).toBeDefined();
  });
  it("should contains 4 input fields", () => {
    const inputs = wrapper.find("input.input");
    expect(inputs.length).toEqual(4);
  });

  it("should contain inputs with names (proposalCode, proposalTitle, observationNight and principalInvestigator)", () => {
    const proposalCode = wrapper.find("input.input").get(0).props.name;
    expect(proposalCode).toEqual("proposalCode");

    const proposalTitle = wrapper.find("input.input").get(2).props.name;
    expect(proposalTitle).toEqual("proposalTitle");

    const observationNight = wrapper.find("input.input").get(3).props.name;
    expect(observationNight).toEqual("observationNight");

    const principalInvestigator = wrapper.find("input.input").get(1).props.name;
    expect(principalInvestigator).toEqual("principalInvestigator");
  });

  it("render correctly", () => {
    expect(
      toJson(
        shallow(<ProposalForm proposal={{ errors: {} }} onChange={onChange} />)
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <ProposalForm
            proposal={{
              errors: {},
              proposalTitle: "hello"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <ProposalForm
            proposal={{
              errors: {
                principalInvestigator: "this is an error",
                proposalCode: "this is an error",
                proposalTitle: "this is an error"
              },
              proposalCode: "Code1"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
  });
});

describe("Proposal form on change", () => {
  const wrapper = mount(
    <ProposalForm proposal={{ errors: {} }} onChange={onChange} />
  );
  it("should find proposal general defined on proposal search form", () => {
    // jest.spyOn(SearchForm.prototype, 'targetChange');
    SearchForm.prototype.generalChange = jest.fn();
    expect(SearchForm.prototype.generalChange).toBeDefined();
  });

  it("should call onChange with any thing", () => {
    const piForm = wrapper.find("input.principal-investigator-input");
    const pi = piForm.find("input");
    pi.simulate("change", {
      target: { value: "hello", name: "principalInvestigator" }
    });
    expect(onChange).toBeCalledWith({
      errors: { principalInvestigator: "" },
      principalInvestigator: "hello"
    });
  });
});
