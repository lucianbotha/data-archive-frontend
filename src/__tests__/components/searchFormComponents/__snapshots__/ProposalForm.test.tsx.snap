// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proposal Form render correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Proposal Code
      </p>
      <InputField
        className="proposal-code-input"
        name="proposalCode"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Principal investigator
      </p>
      <InputField
        className="principal-investigator-input"
        error=""
        name="principalInvestigator"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Proposal title
      </p>
      <InputField
        className="proposal-title-input"
        name="proposalTitle"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Observation night
      </p>
      <InputField
        className="observation-night-input"
        name="observationNight"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Proposal Form render correctly 2`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Proposal Code
      </p>
      <InputField
        className="proposal-code-input"
        name="proposalCode"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Principal investigator
      </p>
      <InputField
        className="principal-investigator-input"
        error=""
        name="principalInvestigator"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Proposal title
      </p>
      <InputField
        className="proposal-title-input"
        name="proposalTitle"
        onChange={[Function]}
        value="hello"
      />
    </styled.div>
    <styled.div>
      <p>
        Observation night
      </p>
      <InputField
        className="observation-night-input"
        name="observationNight"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`Proposal Form render correctly 3`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Proposal Code
      </p>
      <InputField
        className="proposal-code-input"
        error="this is an error"
        name="proposalCode"
        onChange={[Function]}
        value="Code1"
      />
    </styled.div>
    <styled.div>
      <p>
        Principal investigator
      </p>
      <InputField
        className="principal-investigator-input"
        error="this is an error"
        name="principalInvestigator"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Proposal title
      </p>
      <InputField
        className="proposal-title-input"
        error="this is an error"
        name="proposalTitle"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Observation night
      </p>
      <InputField
        className="observation-night-input"
        name="observationNight"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
</Fragment>
`;
